File: vocab-server/scripts/backfill_dict_level.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
backfill_dict_level.py
词典覆盖率历史数据 level 字段分块回填脚本 (VOCAB-Q-PERF-01)
使用 Python 标准库 sqlite3 与 json，跨平台零依赖。
"""

import sys
import os
import json
import time
import sqlite3

def get_db_path():
    if len(sys.argv) > 1:
        return sys.argv[1]
    prod_path = '/var/www/super-agent/vocab.db'
    if os.path.exists(prod_path):
        return prod_path
    local_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), '../vocab.db')
    return local_path

def main():
    db_path = get_db_path()
    print(f"[Backfill Level] 连接数据库: {db_path}")

    if not os.path.exists(db_path):
        print(f"[Backfill Level] 错误: 数据库文件不存在: {db_path}")
        sys.exit(1)

    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()

    # 1. 确保 level 列存在
    cursor.execute("PRAGMA table_info(dict_query_log)")
    columns = [col[1] for col in cursor.fetchall()]
    if 'level' not in columns:
        cursor.execute("ALTER TABLE dict_query_log ADD COLUMN level TEXT")
        conn.commit()
        print("[Backfill Level] 新增 level 列成功。")

    # 2. 确保 (is_success, level) 索引存在
    cursor.execute("CREATE INDEX IF NOT EXISTS idx_dict_log_level ON dict_query_log(is_success, level)")
    conn.commit()
    print("[Backfill Level] 索引 idx_dict_log_level 准备就绪。")

    # 3. 统计待回填数量
    cursor.execute("SELECT COUNT(*) FROM dict_query_log WHERE is_success = 1 AND level IS NULL")
    pending_count = cursor.fetchone()[0]
    print(f"[Backfill Level] 待回填成功记录总数: {pending_count}")

    if pending_count == 0:
        print("[Backfill Level] 没有需要回填的数据，脚本结束。")
        conn.close()
        return

    batch_size = 1000
    processed = 0
    start_time = time.time()

    while True:
        cursor.execute(
            "SELECT id, response_payload FROM dict_query_log WHERE is_success = 1 AND level IS NULL LIMIT ?",
            (batch_size,)
        )
        rows = cursor.fetchall()
        if not rows:
            break

        updates = []
        for row_id, payload_str in rows:
            level = ''
            if payload_str:
                try:
                    parsed = json.loads(payload_str)
                    raw_level = None
                    if isinstance(parsed, dict):
                        if isinstance(parsed.get('payload'), dict):
                            raw_level = parsed['payload'].get('level')
                        if not raw_level:
                            raw_level = parsed.get('level')
                    if isinstance(raw_level, str) and raw_level.strip():
                        level = raw_level.strip()
                except Exception:
                    level = ''
            updates.append((level, row_id))

        cursor.executemany("UPDATE dict_query_log SET level = ? WHERE id = ?", updates)
        conn.commit()
        processed += len(rows)

        elapsed = time.time() - start_time
        percent = (processed / pending_count) * 100
        print(f"[Backfill Level] 进度: {processed}/{pending_count} ({percent:.1f}%) | 耗时: {elapsed:.1f}s")

    conn.close()
    total_time = time.time() - start_time
    print(f"[Backfill Level] 回填完成！共处理 {processed} 行，总耗时: {total_time:.1f}s")

if __name__ == '__main__':
    main()

File: vocab-server/scripts/backfill-dict-level.js
/**
 * backfill-dict-level.js
 * 词典覆盖率历史数据 level 字段分块回填脚本 (VOCAB-Q-PERF-01)
 *
 * 采用分批查询与事务批量更新，避免一次性加载全部 payload 导致内存溢出。
 */

const fs = require('fs');
const path = require('path');
const Database = require('better-sqlite3');

const isProd = process.env.NODE_ENV === 'production' || __dirname.includes('/opt/vocab-server');
const defaultDbPath = isProd ? '/var/www/super-agent/vocab.db' : path.join(__dirname, '../vocab.db');
const dbPath = process.env.DB_PATH || process.argv[2] || defaultDbPath;

console.log(`[Backfill Level] 连接数据库: ${dbPath}`);

if (!fs.existsSync(dbPath)) {
  console.error(`[Backfill Level] 数据库文件不存在: ${dbPath}`);
  process.exit(1);
}

const db = new Database(dbPath);
db.pragma('journal_mode = WAL');

// 1. 确保 level 字段存在
try {
  db.prepare('ALTER TABLE dict_query_log ADD COLUMN level TEXT').run();
  console.log('[Backfill Level] 新增 level 列成功。');
} catch (e) {
  // 列已存在，忽略
}

// 2. 确保 (is_success, level) 索引存在
try {
  db.prepare('CREATE INDEX IF NOT EXISTS idx_dict_log_level ON dict_query_log(is_success, level)').run();
  console.log('[Backfill Level] 索引 idx_dict_log_level 准备就绪。');
} catch (e) {
  console.warn('[Backfill Level] 创建索引跳过:', e.message);
}

// 3. 统计待处理记录数
const pendingCount = db.prepare('SELECT COUNT(*) as count FROM dict_query_log WHERE is_success = 1 AND level IS NULL').get().count;
console.log(`[Backfill Level] 待回填成功记录总数: ${pendingCount}`);

if (pendingCount === 0) {
  console.log('[Backfill Level] 没有需要回填的数据，脚本结束。');
  process.exit(0);
}

const BATCH_SIZE = 1000;
let processed = 0;
let updatedCount = 0;

const selectStmt = db.prepare(`
  SELECT id, response_payload
  FROM dict_query_log
  WHERE is_success = 1 AND level IS NULL
  LIMIT ?
`);

const updateStmt = db.prepare(`
  UPDATE dict_query_log
  SET level = ?
  WHERE id = ?
`);

const updateBatch = db.transaction((rows) => {
  for (const row of rows) {
    let level = '';
    if (row.response_payload) {
      try {
        const parsed = JSON.parse(row.response_payload);
        const raw = parsed?.payload?.level || parsed?.level;
        if (typeof raw === 'string' && raw.trim()) {
          level = raw.trim();
        }
      } catch (e) {
        // payload parse 失败，设为空字符串
        level = '';
      }
    }
    updateStmt.run(level, row.id);
    updatedCount++;
  }
});

const startTime = Date.now();

while (true) {
  const rows = selectStmt.all(BATCH_SIZE);
  if (!rows || rows.length === 0) {
    break;
  }

  updateBatch(rows);
  processed += rows.length;

  const elapsedSec = ((Date.now() - startTime) / 1000).toFixed(1);
  const percent = ((processed / pendingCount) * 100).toFixed(1);
  console.log(`[Backfill Level] 进度: ${processed}/${pendingCount} (${percent}%) | 耗时: ${elapsedSec}s`);
}

console.log(`[Backfill Level] 回填完成！共处理 ${processed} 行，耗时: ${((Date.now() - startTime) / 1000).toFixed(1)}s`);
